File: leet-code/sumOfLeftLeaves.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sumOfLeftLeaves(root *TreeNode) int {
	return reverseTree(root)
}

func reverseTree(root *TreeNode) int {
	ans := 0
	if root == nil {
		return ans
	}
	if root.Left != nil {
		if ok := isLeaf(root.Left); ok {
			ans += root.Left.Val
		} else {
			ans += reverseTree(root.Left)
		}
	}

	if ok := isLeaf(root.Right); root.Right != nil && !ok {
		ans += reverseTree(root.Right)
	}
	return ans
}

func isLeaf(root *TreeNode) bool {
	if root == nil {
		return false
	}
	if root.Left == nil && root.Right == nil {
		return true
	}
	return false
}

File: leet-code/sumOfLeftLeaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumOfLeftLeaves = function (root) {
  let reverse = function (root) {
    let ans = 0

    if (root == null) {
      return ans
    }
    if (root.left) {
      ans += isLeafNode(root.left) ? root.left.val : reverse(root.left)
    }
    if (root.right && !isLeafNode(root.right)) {
      ans += reverse(root.right)
    }
    return ans
  }

  let isLeafNode = function (node) {
    return node.left == null && node.right == null
  }

  return reverse(root)
}
